internal/provider/resource_cloudflare_account.go

```diff
@@ -0,0 +1,130 @@
+)
+
+func resourceCloudflareAccount() *schema.Resource {
+	return &schema.Resource{
+		Schema:        resourceCloudflareAccountSchema(),
+		CreateContext: resourceCloudflareAccountCreate,
+		ReadContext:   resourceCloudflareAccountRead,
+		UpdateContext: resourceCloudflareAccountUpdate,
+		DeleteContext: resourceCloudflareAccountDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Description: heredoc.Doc(`
+			Provides a Cloudflare Account resource. Account is the basic resource for
+			working with Cloudflare zones, teams and users. 
+		`),
+	}
+}
+
+func resourceCloudflareAccountCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*cloudflare.API)
+	accountName := d.Get("name").(string)
+	accountType := d.Get("type").(string)
+
+	tflog.Debug(ctx, fmt.Sprintf("Creating Cloudflare Account: name %s", accountName))
+
+	account := cloudflare.Account{
+		Name: accountName,
+		Type: accountType,
+	}
+	acc, err := client.CreateAccount(ctx, account)
+
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("error creating account %q: %w", accountName, err))
+	}
+
+	d.SetId(acc.ID)
+
+	return resourceCloudflareAccountRead(ctx, d, meta)
+}
+
+func resourceCloudflareAccountRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*cloudflare.API)
+	accountID := d.Id()
+
+	foundAcc, _, err := client.Account(ctx, accountID)
+	if err != nil || foundAcc.ID == "" {
+		var notFoundError *cloudflare.NotFoundError
+		if errors.As(err, &notFoundError) {
+			tflog.Info(ctx, fmt.Sprintf("Account %s no longer exists", d.Id()))
+			d.SetId("")
+			return nil
+		}
+		return diag.FromErr(fmt.Errorf("error finding Account %q: %w", d.Id(), err))
+	}
+
+	tflog.Debug(ctx, fmt.Sprintf("AccountDetails: %#v", foundAcc))
+
+	d.Set("name", foundAcc.Name)
+	d.Set("type", foundAcc.Type)
+	d.Set("enforce_twofactor", foundAcc.Settings.EnforceTwoFactor)
+
+	return nil
+}
+
+func resourceCloudflareAccountUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*cloudflare.API)
+	accountID := d.Id()
+	foundAcc, _, err := client.Account(ctx, accountID)
+	if err != nil || foundAcc.ID == "" {
+		return diag.FromErr(fmt.Errorf("error finding Account %q: %w", d.Id(), err))
+	}
+
+	tflog.Debug(ctx, fmt.Sprintf("Updating Cloudflare Account: id %s", accountID))
+
+	if accountName, ok := d.GetOk("name"); ok && d.HasChange("name") {
+		foundAcc.Name = accountName.(string)
+	}
+
+	if enforce_twofactor, ok := d.GetOk("enforce_twofactor"); ok && d.HasChange("enforce_twofactor") {
+		foundAcc.Settings.EnforceTwoFactor = enforce_twofactor.(bool)
+	}
+
+	_, err = client.UpdateAccount(ctx, accountID, foundAcc)
+	if err != nil {
+		var cfError *cloudflare.RequestError
+		if errors.As(err, &cfError) {
+			errCodes := cfError.ErrorCodes()
+			if len(errCodes) == 1 && errCodes[0] == 1001 {
```

Comment:
can you elaborate on this issue? with the correct entitlements, this should work.

Reply:
This is kind of a workaround to maybe something that should be fixed in cloudflare-go library. For example, running such basic code will actually make the update to work but will raise a 1001 error on `err3`:
```
	acc, _, err2 := api.Account(ctx, someAccountID)
	if err2 != nil {
		log.Fatal(err2)
	}

	fmt.Println(acc.ID)
	acc.Name = acc.Name + "x"
	_, err3 := api.UpdateAccount(ctx, acc.ID, acc)
	if err3 != nil {
		log.Fatal(err3)
	}
```

However, the update of the name will actually succeed. I think a change would be needed in https://github.com/cloudflare/cloudflare-go/blob/master/accounts.go#L95 so that accountType is not sent to backend while updating

Reply:
I will make this handling more clean by setting Type to empty as the JSON will omit it if empty on update

Reply:
https://github.com/cloudflare/terraform-provider-cloudflare/pull/1902#:~:text=prevent%201001%20error-,bdaf421,-Merge%20state will make sure the error does not pop up by omitting Type in the update.